packages/apollo-engine-reporting/src/schemaReporter.ts

```diff
@@ -0,0 +1,166 @@
+export class SchemaReporter {
+  // These mirror the gql variables
+  private readonly serverInfo: EdgeServerInfo;
+  private readonly executableSchemaDocument: any;
+  private readonly url: string;
+
+  private isStopped: boolean;
+  private readonly headers: Headers;
+
+  constructor(
+    serverInfo: EdgeServerInfo,
+    schemaSdl: string,
+    apiKey: string,
+    schemaReportingEndpoint: string | undefined,
+  ) {
+    this.headers = new Headers();
+    this.headers.set('Content-Type', 'application/json');
+    this.headers.set('x-api-key', apiKey);
+    this.headers.set('apollographql-client-name', 'apollo-engine-reporting');
+    this.headers.set(
+      'apollographql-client-version',
+      require('../package.json').version,
+    );
+
+    this.url =
+      schemaReportingEndpoint ||
+      'https://engine-graphql.apollographql.com/api/graphql';
+
+    this.serverInfo = serverInfo;
+    this.executableSchemaDocument = schemaSdl;
+    this.isStopped = false;
+  }
+
+  public stopped(): Boolean {
+    return this.isStopped;
+  }
+
+  public stop() {
+    this.isStopped = true;
+  }
+
+  public async reportServerInfo(
+    withExecutableSchema: boolean,
+  ): Promise<ReportServerInfoReturnVal> {
+    const { data, errors } = await this.graphManagerQuery({
+      info: this.serverInfo,
+      executableSchema: withExecutableSchema
+        ? this.executableSchemaDocument
+        : null,
+    } as ReportServerInfoVariables);
+
+    if (errors) {
+      throw new Error((errors || []).map((x: any) => x.message).join('\n'));
+    }
+
+    if (!data || !data.me || !data.me.__typename) {
+      throw new Error(`
+Heartbeat response error. Received incomplete data from Apollo Graph Manager.
+If this continues please reach out at support@apollographql.com.
+Got response: "${JSON.stringify(data)}"
+      `);
+    }
+
+    if (data.me.__typename === 'UserMutation') {
+      this.isStopped = true;
+      throw new Error(`
+      This server was configured with an API key for a user.  Only a service's API key may be used for schema reporting.
+      Please visit the settings for this graph at https://engine.apollographql.com/ to obtain an API key for a service.
+      `);
+    } else if (data.me.__typename === 'ServiceMutation') {
+      if (!data.me.reportServerInfo) {
+        throw new Error(`
+Heartbeat response error. Received incomplete data from Apollo Graph Manager.
+If this continues please reach out at support@apollographql.com.
+Got response: "${JSON.stringify(data)}"
+      `);
+      }
+      return data.me.reportServerInfo;
+    } else {
+      throw new Error(`
+Unexpected response. Received unexpected data from Apollo Graph Manager
+If this continues please reach out at support@apollographql.com.
+Got response: "${JSON.stringify(data)}"
+      `);
+    }
+  }
+
+  private async graphManagerQuery(
+    variables: ReportServerInfoVariables,
+  ): Promise<AutoregReportServerInfoResult> {
```

Comment:
When I first noticed the inconsistency between this type name (which is prefixed with `Autoreg`) and the others (which lack that prefix), I was tempted to suggest aligning them, but then I realized they were auto-generated.  I'm not sure how to resolve that, and I'm not asking you, but just thought I'd note that perspective of mine.

Reply:
Renamed the types to get rid of `Autoreg`